feat: added SocialRemark util (#136)
dappforce/subsocial-js: packages/utils/src/socialRemark/README.md

# SocialRemark

> SocialRemark protocol is designed to process cross-chain actions.

Reviewer: ```suggestion
The SocialRemark protocol is designed to process cross-chain actions.
```

## SocialRemark message format

- Domain Registration
  `prot_name::version::action::op_id::target::domain_name::token`
- Energy Generation
  `prot_name::version::action::op_id::target::energy_amount::token`

> Remark message has 2 parts:

Reviewer: ```suggestion
A Remark message has 2 parts:
```

> - core part (is the same for all messages): `prot_name::version::action`

Reviewer: ```suggestion
- the core part (which is the same for all messages): `prot_name::version::action`
```

> - content part (depends on actions):

Reviewer: ```suggestion
- the content part (which depends on the action):
```

  - `op_id::target::domain_name::token`
  - `op_id::target::energy_amount::token`

Core part options:

> - `prot_name` - protocol name (default valid value `social` but can be configured but `.setConfig()` method)

Reviewer: ```suggestion
- `prot_name` - The protocol name (the default valid value is `social`, but it can be configured with the `.setConfig()` method)
```

> - `version` - protocol version (current available version `0.1`)

Reviewer: ```suggestion
- `version` - The protocol version (the current available version is `0.1`)
```

> - `action` - message action (available actions: `DMN_REG`, `DMN_REG_OK`, `DMN_REG_REFUND`, `NRG_GEN`, `NRG_GEN_OK`, `NRG_GEN_REFUND`)

Reviewer: ```suggestion
- `action` - The message action (available actions: `DMN_REG`, `DMN_REG_OK`, `DMN_REG_REFUND`, `NRG_GEN`, `NRG_GEN_OK`, `NRG_GEN_REFUND`)
```

Content options:

> - `op_id` - unique operation identified. Should be the same for all actions of one operation (the same `op_id` value for `DMN_REG` , `DMN_REG_OK` , `DMN_REG_REFUND` of A single domain registration flow. Necessary for relating actions with each other.)

Reviewer: ```suggestion
- `op_id` - A unique operation identifier. It should be the same for all actions of one operation (the same `op_id` value for `DMN_REG` , `DMN_REG_OK` , `DMN_REG_REFUND` of A single domain registration flow. Necessary for relating actions with each other.)
```
I'm not really understanding this one @mckrava, can you please try to rephrase it and then ping me?

Author: `op_id` - A unique operation identifier. It should have the same value for all actions of one operation/action (the same `op_id` value for  one chain of Domain Registration actions `DMN_REG` -> `DMN_REG_OK` || `DMN_REG_REFUND`  It's necessary to build relationship between actions of a single purchase order.)

@Yung-Beef

Reviewer: ```suggestion
`op_id` - A unique operation identifier. It should have the same value for all instances of a particular type of operation/action (the same `op_id` value for one chain of Domain Registration actions `DMN_REG` -> `DMN_REG_OK` || `DMN_REG_REFUND` It's necessary to build relationship between actions of a single purchase order.)
```
The first part is good now, but the stuff in the parentheses is confusing me... the first part says to use the same value for all instances of a particular type of operation, but then there are 3 operations mentioned in the parentheses?

Author: Ah, ok. Let me explain an idea - we have such action like `Domain Generation Request` (`DMN_REG`). UI app (Polkaverse) creates registration request (system remark) and UI app adds unique ID (`op_id === 0x123`) for this particular request.  Then our seller-squid catches this request and creates `DomainRegistrationOrder` with ID === `op_id` === 0x123. Then, if registration is successfully done by seller-squid, it sends to Polkadot chain another remark as confirmation of successful registration (`DMN_REG_OK`) or refund (`DMN_REG_REFUND`) request. Both of these requests/remarks must have op_id === 0x123  to understand that this particular confirmation or refund request belongs to specific DomainRegistrationOrder with ID === 0x123. 
I hope it should help to compile human-readable explanation )

@Yung-Beef

Reviewer: ```suggestion
`op_id` - A unique operation identifier that is used to build a relationship between Domain Registration actions, in order to associate a particular `DMN_REG_OK` or `DMN_REG_REFUND` to the corresponding `DMN_REG`.
```

How does that look? 

If the || means "or", you can switch it back to || if you think it would be easier for devs to read.

> - `target` - recipient of domain or energy _(address in any SS58 format, SocialRemark instance will convert it to SubSocial format underhood (prefix: 28))_

Reviewer: ```suggestion
- `target` - The recipient of a domain or energy _(this can be an address in any SS58 format, the SocialRemark instance will convert it to Subsocial's format under the hood (prefix: 28))_
```
"recipient of a domain" implies that a domain is being transferred. Is that the only time you would use `target`?

Author: @Yung-Beef

Reviewer: ```suggestion
- `target` - The receiver of a domain or recipient of energy _(this can be an address in any SS58 format, the SocialRemark instance will convert it to Subsocial's format under the hood (prefix: 28))_
```
Or should it be "the owner of a new domain"?

Author: `receiver`  is fine I guess

> - `domain_name` - domain name purchased by target

Reviewer: ```suggestion
- `domain_name` - The domain name purchased by the target
```

> - `energy_amount` - amount of energy purchased by target

Reviewer: ```suggestion
- `energy_amount` - The amount of energy purchased by the target
```

> - `token` - used token for purchase process

Reviewer: ```suggestion
- `token` - The type of token (for example, SUB, DOT, or KSM) used in the purchase process
```

> Dummy example of SocialRemark message:

Reviewer: ```suggestion
Here is a dummy example of a SocialRemark message:
```

```
social::0.1::DMN_REG::0x44d8d9f1bc70e45eb773731f9ffc5d3646df56497c40cdfff37c8ceb71fa2-2104480009442407::3t5NA8UKsGzrCDMfp8XMEBghiYthWGXGsHbjtJY45NUJDY5P::somenewdomain.sub::DOT
```

## Usage examples:

```typescript
import { SocialRemark, SubSclSource } from '@subsocial-js/utils'
import { randomAsNumber } from '@polkadot/util-crypto'

/**

>  * Set custom protocol name(s), which will be recognized as valid protocol name.

Reviewer: ```suggestion
 * Set a custom protocol name(s), which will be recognized as a valid protocol name.
```

>  * Default value - `social` (IMPORTANT - not recommended to use default value

Reviewer: ```suggestion
 * The default value is `social` (IMPORTANT - it is not recommended to use the default value
```

 * for development or testing purposes).
 */
SocialRemark.setConfig({ protNames: ['social_custom'] })

const remarkSource: SubSclSource<'DMN_REG'> = {
  protName: 'social_custom',
  action: 'DMN_REG',
  version: '0.1',
  content: {
    opId: `${randomAsNumber()}`,
    domainName: `somenewdomain.sub`,
    target: '3t5NA8UKsGzrCDMfp8XMEBghiYthWGXGsHbjtJY45NUJDY5P',
    token: 'ROC'
  }
}

/**

>  * Create SocialRemark instance for further usage by calling instance methods.

Reviewer: ```suggestion
 * Creates a SocialRemark instance for further usage by calling the instance methods.
```
the instance methods or these instance methods?

 */
const socialRemarkInstance: SocialRemark = new SocialRemark().fromSource(
  remarkSource
)

/**

>  * Get compiled string remark message for sending to blockchain

Reviewer: ```suggestion
 * Gets a compiled string remark message for sending to the blockchain.
```

 */
const socialRemarkMessageStr: string = socialRemarkInstance.toMessage()

/**

>  * Parse string social remark message and create SocialRemark instance.

Reviewer: ```suggestion
 * Parses a string social remark message and create a SocialRemark instance.
```

Reviewer: Or should it be "the string social remark message"?

Reviewer: ```suggestion
 * Parses the string social remark message and create a SocialRemark instance.
```

 */
const socialRemarkParsed: SocialRemark = new SocialRemark().fromMessage(
  socialRemarkMessageStr
)

/**

>  * Check if remark message is valid SocialRemarl message.

Reviewer: ```suggestion
 * Checks if a remark message is a valid SocialRemark message.
```

 */

> if (!remark.isValidMessage) throw new Error('SocialRemar message is not valid.')

Reviewer: ```suggestion
if (!remark.isValidMessage) throw new Error('SocialRemark message is not valid.')
```
